Sources/CoreMetrics/Metrics.swift

```diff
@@ -12,220 +12,230 @@
 //
 //===----------------------------------------------------------------------===//
 
-public protocol Counter: AnyObject {
+public protocol CounterHandler: AnyObject {
     func increment<DataType: BinaryInteger>(_ value: DataType)
 }
 
-public extension Counter {
+public class Counter: CounterHandler {
+    @usableFromInline
+    var handler: CounterHandler
+    public let label: String
+    public let dimensions: [(String, String)]
+
+    // this method is public to provide an escape hatch for situations one must use a custom factory instead of the gloabl one
+    // we do not expect this API to be used in normal circumstances, so if you find yourself using it make sure its for a good reason
+    public init(label: String, dimensions: [(String, String)], handler: CounterHandler) {
+        self.label = label
+        self.dimensions = dimensions
+        self.handler = handler
+    }
+
     @inlinable
-    func increment() {
+    public func increment<DataType: BinaryInteger>(_ value: DataType) {
+        self.handler.increment(value)
+    }
+
+    @inlinable
+    public func increment() {
         self.increment(1)
     }
 }
 
-public protocol Recorder: AnyObject {
+public extension Counter {
+    public convenience init(label: String, dimensions: [(String, String)] = []) {
+        let handler = MetricsSystem.handler.makeCounter(label: label, dimensions: dimensions)
+        self.init(label: label, dimensions: dimensions, handler: handler)
+    }
+}
+
+public extension Counter {
+    @inlinable
+    public static func `do`(label: String, dimensions: [(String, String)] = [], body: (Counter) -> Void) {
```

Comment:
A bit unsure about the `do` notation... Did I miss some discussion which led us to go with the `do` style? 
I'm not really use it leans to cleaner code and it also leads to repeating the string labels all the time....

Separate questions here really:
- why the move to `do` from the `with...` style... To allow swapping the underlying type more easily?  Not sure if not "seeing on call site" which kind of metric is recoded is a feature.
- such API seems to be easy to add by users by an extension right? So are we sure we have to ship this in the minimal API proposal?

Reply:
1/ the ‘with’ variants were on ‘Metrics’ and read fluently, eg ‘Metrics.withCounter’. as we now only use ‘Metrics’ (now called ‘MetricsSystem’) to bootstrap, moved these to the relevant class, eg ‘Counter.do’ which imo reads better than ‘Counter.with’. open to better ideas tho

2/ fair point, we could just drop these as they are syntactic auger. would love to hear what others think

Reply:
1. I'm okay with `do` as it does mirror Swift's statement for introducing a block, but in that case, shouldn't it be `rethrows`?

2. I haven't used a metrics framework before, so I'm unsure how important/frequent it is to want to use a `Counter` like this.

Reply:
After thinking about it some more, I'm not even sure the syntactic sugar of `do` is even worth it anymore:

```swift
Counter.do(label: "test") { counter
    // use counter
}

// VS

let counter = Counter(label: "test")
// use counter
```

Reply:
okay let me get rid of that

Reply:
https://github.com/tomerd/swift-server-metrics-api-proposal/pull/9/commits/84e8ad29a41e73586f20437c7ac0023fb9ef1a7a